NoClassDefFoundError in IntelliJ plugin while trying to generate test

**Description**

NoClassDefFoundError in IntelliJ plugin trying to generate test

**To Reproduce**

Steps to reproduce the behavior:

_Example:_

1. Run the 'javaparser' project in IntelliJ Idea
2. Create file with the content:
```
package com.github.javaparser;

import com.github.javaparser.ast.CompilationUnit;

public class Parser {
    ParseResult<CompilationUnit> parse(String text) {
        return new JavaParser().parse(text);
    }
}

```
3. Use plugin to generate tests for Parser

**Expected behavior**

Tests are supposed to be generated.

**Actual behavior**

Exception is thrown

**Visual proofs (screenshots, logs, images)**
```
com.jetbrains.rd.util.reactive.RdFault: com/google/common/cache/CacheLoader, reason: java.lang.NoClassDefFoundError: com/google/common/cache/CacheLoader
	at soot.Scene.v(Scene.java:178)
	at org.utbot.framework.util.SootUtilsKt.addBasicClasses(SootUtils.kt:137)
	at org.utbot.framework.util.SootUtilsKt.initSoot(SootUtils.kt:89)
	at org.utbot.framework.util.SootUtilsKt.access$initSoot(SootUtils.kt:1)
	at org.utbot.framework.util.SootUtils.runSoot(SootUtils.kt:46)
	at org.utbot.framework.plugin.api.TestCaseGenerator.<init>(TestCaseGenerator.kt:88)
	at org.utbot.framework.plugin.api.TestCaseGenerator.<init>(TestCaseGenerator.kt:61)
	at org.utbot.framework.process.EngineMainKt$setup$2.invoke(EngineMain.kt:84)
	at org.utbot.framework.process.EngineMainKt$setup$2.invoke(EngineMain.kt:81)
	at org.utbot.rd.CallsSynchronizer$measureExecutionForTermination$1$1.invoke(ClientProcessUtil.kt:84)
	at org.utbot.rd.CallsSynchronizer.measureExecutionForTermination(ClientProcessUtil.kt:75)
	at org.utbot.rd.CallsSynchronizer$measureExecutionForTermination$1.invoke(ClientProcessUtil.kt:83)
	at com.jetbrains.rd.framework.IRdEndpoint$set$1.invoke(TaskInterfaces.kt:182)
	at com.jetbrains.rd.framework.IRdEndpoint$set$1.invoke(TaskInterfaces.kt:182)
	at com.jetbrains.rd.framework.impl.RdCall.onWireReceived(RdTask.kt:360)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2$2.invoke(MessageBroker.kt:57)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2$2.invoke(MessageBroker.kt:56)
	at com.jetbrains.rd.framework.impl.ProtocolContexts.readMessageContextAndInvoke(ProtocolContexts.kt:148)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2.invoke(MessageBroker.kt:56)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2.invoke(MessageBroker.kt:54)
	at com.jetbrains.rd.util.threading.SingleThreadSchedulerBase.queue$lambda-3(SingleThreadScheduler.kt:41)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassNotFoundException: com.google.common.cache.CacheLoader
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 24 more

	at com.jetbrains.rd.framework.RdTaskResult$Companion.read(TaskInterfaces.kt:30)
	at com.jetbrains.rd.framework.impl.CallSiteWiredRdTask.onWireReceived(RdTask.kt:104)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2$2.invoke(MessageBroker.kt:57)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2$2.invoke(MessageBroker.kt:56)
	at com.jetbrains.rd.framework.impl.ProtocolContexts.readMessageContextAndInvoke(ProtocolContexts.kt:148)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2.invoke(MessageBroker.kt:56)
	at com.jetbrains.rd.framework.MessageBroker$invoke$2.invoke(MessageBroker.kt:54)
	at com.jetbrains.rd.framework.impl.RdCall$createResponseScheduler$1.queue$execute(RdTask.kt:278)
	at com.jetbrains.rd.framework.impl.RdCall$createResponseScheduler$1.access$queue$execute(RdTask.kt:267)
	at com.jetbrains.rd.framework.impl.RdCall$createResponseScheduler$1$queue$1.invokeSuspend(RdTask.kt:287)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at org.utbot.intellij.plugin.process.EngineProcess.createTestGenerator(EngineProcess.kt:190)
	at org.utbot.intellij.plugin.generator.UtTestsDialogProcessor$createTests$1$1.run(UtTestsDialogProcessor.kt:158)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$5(CoreProgressManager.java:493)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:244)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:189)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:176)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:244)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:829)

```
![image](https://user-images.githubusercontent.com/11944514/200824538-1e2ba0b4-2585-41d6-89e4-7ed462dbf224.png)
~~~

**Environment**

IntelliJ IDEA 2022.1.4 (Community Edition)
Build #IC-221.6008.13, built on July 19, 2022
Runtime version: 11.0.15+10-b2043.56 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Linux 5.15.0-50-generic
GC: G1 Young Generation, G1 Old Generation
Memory: 2048M
Cores: 20
Non-Bundled Plugins:
    org.jetbrains.kotlin (221-1.7.21-release-272-IJ5591.52)
    org.utbot.intellij.plugin.id (2022.11)

Kotlin: 221-1.7.21-release-272-IJ5591.52
Current Desktop: ubuntu:GNOME

--- a/utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt
+++ b/utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt
@@ -614,6 +614,11 @@ class UtLambdaModel(
     companion object {
         private const val fakeName = "<FAKE>"
 
+        /**
+         * Create a non-existent lambda with fake method.
+         *
+         * That's temporary solution for building lambdas from concrete values.
+         */
         fun createFake(id: Int, samType: ClassId, declaringClass: ClassId) =
             UtLambdaModel(id, samType, declaringClass, fakeName)
     }

--- a/utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt
+++ b/utbot-framework-api/src/main/kotlin/org/utbot/framework/plugin/api/Api.kt
@@ -570,12 +570,31 @@ class UtLambdaModel(
     val capturedValues: MutableList<UtModel> = mutableListOf(),
 ) : UtReferenceModel(id, samType) {
 
+    val isFake: Boolean = lambdaName == fakeName
+
     val lambdaMethodId: MethodId
-        get() = declaringClass.jClass
-            .declaredMethods
-            .singleOrNull { it.name == lambdaName }
-            ?.executableId // synthetic lambda methods should not have overloads, so we always expect there to be only one method with the given name
-            ?: error("More than one method with name $lambdaName found in class: ${declaringClass.canonicalName}")
+        get() {
+            if (isFake) {
+                val targetMethod = samType.jClass.declaredMethods.single()
+                return object : MethodId(
+                    declaringClass,
+                    fakeName,
+                    targetMethod.returnType.id,
+                    targetMethod.parameterTypes.map { it.id }
+                ) {
+                    override val modifiers: Int = ModifierFactory.invoke {
+                        public = true
+                        static = true
+                        final = true
+                    }
+                }
+            }
+            return declaringClass.jClass
+                .declaredMethods
+                .singleOrNull { it.name == lambdaName }
+                ?.executableId // synthetic lambda methods should not have overloads, so we always expect there to be only one method with the given name
+                ?: error("More than one method with name $lambdaName found in class: ${declaringClass.canonicalName}")
+        }
 
     override fun toString(): String = "Anonymous function $lambdaName implementing functional interface $declaringClass"
 
@@ -591,6 +610,13 @@ class UtLambdaModel(
     }
 
     override fun hashCode(): Int = id
+
+    companion object {
+        private const val fakeName = "<FAKE>"
+
+        fun createFake(id: Int, samType: ClassId, declaringClass: ClassId) =
+            UtLambdaModel(id, samType, declaringClass, fakeName)
+    }
 }
 
 /**

--- a/utbot-framework/src/main/kotlin/org/utbot/framework/concrete/UtModelConstructor.kt
+++ b/utbot-framework/src/main/kotlin/org/utbot/framework/concrete/UtModelConstructor.kt
@@ -35,6 +35,7 @@ import org.utbot.framework.util.valueToClassId
 import java.lang.reflect.Modifier
 import java.util.IdentityHashMap
 import java.util.stream.BaseStream
+import org.utbot.framework.plugin.api.util.utContext
 
 /**
  * Represents common interface for model constructors.
@@ -79,6 +80,21 @@ class UtModelConstructor(
         return objectToModelCache[value]?.let { (it as? UtReferenceModel)?.id } ?: computeUnusedIdAndUpdate()
     }
 
+    private val proxyLambdaSubstring = "$\$Lambda$"
+
+    private fun isProxyLambda(value: Any?): Boolean {
+        if (value == null) {
+            return false
+        }
+        return proxyLambdaSubstring in value::class.java.name
+    }
+
+    private fun constructFakeLambda(value: Any, classId: ClassId): UtLambdaModel {
+        val baseClassName = value::class.java.name.substringBefore(proxyLambdaSubstring)
+        val baseClass = utContext.classLoader.loadClass(baseClassName).id
+        return UtLambdaModel.createFake(handleId(value), classId, baseClass)
+    }
+
     /**
      * Constructs a UtModel from a concrete [value] with a specific [classId]. The result can be a [UtAssembleModel]
      * as well.
@@ -90,6 +106,9 @@ class UtModelConstructor(
             if (model is UtLambdaModel) {
                 return model
             }
+        }
+        if (isProxyLambda(value)) {
+            return constructFakeLambda(value!!, classId)
         }
         return when (value) {
             null -> UtNullModel(classId)

